### ISSUE TITLE
Modify DFT grid schemes

### ISSUE THREAD

**Issue by KarimAED** [pyscf/pyscf#2178]
The DFT grid schemes currently exhibit a few behaviours that seem odd / could be implemented with improvements:

1. The Treutler radial grids are implemented with different parametrization for $\xi$ to the one suggested in Treutler's original paper.
2. Different radial shells [are shuffled](https://github.com/pyscf/pyscf/blob/master/pyscf/dft/gen_grid.py#L230) in `gen_atomic_grids`. This has to then be compensated for with the `sort_grids` kwarg.
3. This kwarg in turn **cannot** be overridden in calls to the top-level API, making this routine inflexible.

Point 1 is simply a divergence from the described optimal behaviour in the paper, which I believe should be fixed.

Points 2 & 3 constitute an issue for our work, as we are trying to validate our own code (DQC) against PySCF. However, there is no way of aligning the grids without outright sorting them for both codes or changing the way PySCF approaches this. As the former is expensive to do repeatedly, and I believe that the PySCF routine is currently suboptimal, I suggest the latter.

I am happy to contribute a PR addressing these points.

**Comment by susilehtola** [comment]
Per 1, do you mean that PySCF uses different values for the atomic radii?

**Comment by KarimAED** [comment]
No, I don't think that is the issue. If you have a look at [the Treutler paper](https://pubs.aip.org/aip/jcp/article-abstract/102/1/346/482489/Efficient-molecular-numerical-integration-schemes?redirectedFrom=fulltext), it parametrizes the M4 radial transform (eq. 19) in terms of $\xi$, a value that is optimized on a per-element basis in a table further down in the paper. In DQC, these values are implemented as seen in [the code](https://github.com/OxfordHED/dqc/blob/master/dqc/grid/factory.py#L175). However, to obtain the same number of grid points as I get with PySCF, this needs to be overwritten with $\xi = 1$ in DQC. The resulting grids then contain the same points (up to padding and sorting).

The $\xi$ factor changes the spacing relative to the atomic radii values, thereby changing how many radial points correspond to a given Lebedev order in the NWChem truncation.

**Comment by KarimAED** [comment]
I can also verify that the code in [this function](https://github.com/pyscf/pyscf/blob/master/pyscf/dft/radi.py#L99) in PySCF corresponds to $\xi = 1$. However, I am not sure whether the code [here](https://github.com/pyscf/pyscf/blob/master/pyscf/dft/radi.py#L139) is supposed to change this value after the fact. If it is, something seems to be off, as per my comparison with DQC.

**Comment by susilehtola** [comment]
> No, I don't think that is the issue. If you have a look at [the Treutler paper](https://pubs.aip.org/aip/jcp/article-abstract/102/1/346/482489/Efficient-molecular-numerical-integration-schemes?redirectedFrom=fulltext), it parametrizes the M4 radial transform (eq. 19) in terms of ξ, a value that is optimized on a per-element basis in a table further down in the paper. 

The $\xi$ parameters in the Treutler-Ahlrichs paper are none other than effective atomic radii: they write "The scaling parameters $\xi$ have first been optimized in isolated atom calculations, the resulting values have then been checked in molecular calculations which lead to essential changes for the hydrogen atom only."

Similar scalings are also employed in many other grid schemes, and typical DFT codes separate the application of the atomic radius from the generation of the radial quadrature rule to eliminate code duplication.

**Comment by KarimAED** [comment]
Ahh, apologies, I wasn't aware of this. However, the values for $\xi$ in the paper are different from the atomic radii found [here](https://github.com/pyscf/pyscf/blob/master/pyscf/data/radii.py#L23) though, right? In addition, that still doesn't explain why I see a discrepancy which disappears only when I set the atomic radii ($\xi$) all to 1 in DQC. Here is the PySCF grid generation I am using:

```py
    mol_pyscf = gto.M(atom=moldesc, basis="pc-2", spin=spin, unit="B")
    pyscf_dft = dft.KS(mol_pyscf, "lda")
    pyscf_dft.small_rho_cutoff = 1e-21
    pyscf_dft.grids.alignment = 0
    pyscf_dft.kernel()
    pyscf_grid = pyscf_dft.grids.coords
```

Note that I have set alignment and rho_cutoff to zero to avoid other discrepancies in the grid size. Am I missing something? If I run this against DQC calculations, which use the values for $\xi$ from the paper, the DQC calculations always return slightly larger grids, e.g. 13722 points for DQC, and 12246 points for PySCF. This discrepancy vanishes the moment I set $\xi=1$ in DQC.

Note: I am using a version of DQC that has had additional bugfixes c.f. the version on PyPI. Beforehand, there was also an error where the chosen Lebedev order was off by one.

**Comment by KarimAED** [comment]
With regard to the grid ordering, though, are you happy for me to submit a PR @susilehtola? The idea would be to retain the order of radial shells, rather than using `sorted(set(angs))` in the relevant script. This may create a minimal increase in computational cost, but it should not be significant.

**Comment by susilehtola** [comment]
You need to disable pruning to compare to other codes.
```
 pyscf_dft.grids.prune = None
```

**Comment by susilehtola** [comment]
> With regard to the grid ordering, though, are you happy for me to submit a PR @susilehtola? The idea would be to retain the order of radial shells, rather than using `sorted(set(angs))` in the relevant script. This may create a minimal increase in computational cost, but it should not be significant.

Well I'm not the person who wrote that code ;)
The general rule in these cases is to submit a PR and then have comments in...

**Comment by KarimAED** [comment]
I am actually comparing with both codes using NWChem pruning. It uses the same Lebedev orders now, but how many radial points are associated with each lebedev order differs between the two codes. I can deactivate the discrepancy by changing the $\xi$ values to 1 in DQC.

**Comment by tberkel** [comment]
I assume @sunqm was the author of that part of the code, so maybe he should weigh-in before you write a PR.

**Comment by susilehtola** [comment]
> I am actually comparing with both codes using NWChem pruning. It uses the same Lebedev orders now, but how many radial points are associated with each lebedev order differs between the two codes. I can deactivate the discrepancy by changing the ξ values to 1 in DQC.

NWChem pruning also depends on the employed atomic radii. Are you sure the radii are consistent between these two routines in both PySCF and DQC?

**Comment by KarimAED** [comment]
Yes, I have checked this. The radii in DQC are directly copied from PySCF. For confirmation, compare [PySCF radii](https://github.com/pyscf/pyscf/blob/master/pyscf/data/radii.py#L23) and [DQC radii](https://github.com/OxfordHED/dqc/blob/master/dqc/utils/periodictable.py#L126). Again, the only explanation I can see is that the $\xi$ values are somehow not applied in PySCF.

**Comment by susilehtola** [comment]
Compare the generated grids without pruning. Are they the same?

**Comment by KarimAED** [comment]
Ahh thanks, with pruning turned off, I recover the same length grid. Looks like there is a bug in DQCs NWChem pruning then... Is there a place where the actual pruning scheme is written up in a paper? Because it's rather cumbersome to have only the implementation in PySCF for reference. At any rate, I have actually found a workaround for the other issue I had as well, so there really is no need for this issue to stay open. Thanks for the support @susilehtola and @tberkel!

**Comment by sunqm** [comment]
I don't remember where I found the pruning scheme at beginning. Based on the name of the function, one is taken from NWChem. The other schemes may be taken from Psi3, or BDF code.

**Comment by sunqm** [comment]
> I am actually comparing with both codes using NWChem pruning. It uses the same Lebedev orders now, but how many radial points are associated with each lebedev order differs between the two codes. I can deactivate the discrepancy by changing the ξ values to 1 in DQC.

I guess I overlooked the $\xi$ parameter when I implement the code for some reasons. I think it makes sense to add the missing $\xi$ values to the code IMO. However, updating the radial grids scheme may slightly change the default DFT results. We can first include them and evaluate the impact before setting them to the default.

**Comment by KarimAED** [comment]
Hmm, so I'm not sure whether you necessarily missed something, @sunqm . Once I deactivate NWChem pruning, the two codes give the same result. That would mean one of the following:

1. DQC applies the $\xi$ parameters twice ONLY when NWChem pruning is activated
2. PySCF does not apply $\xi$ parameters ONLY when NWChem pruning is activated

However, I don't believe that the issue necessarily lies with PySCF (DQC is definitely the less reliable code overall), so if you could maybe check whether the $\xi$ values are applied at any point (depending on pruning or no pruning), that would be useful. I am working on the same checks for DQC.

**Comment by sunqm** [comment]
> Hmm, so I'm not sure whether you necessarily missed something, @sunqm . Once I deactivate NWChem pruning, the two codes give the same result. That would mean one of the following:
> 
> 1. DQC applies the ξ parameters twice ONLY when NWChem pruning is activated
> 2. PySCF does not apply ξ parameters ONLY when NWChem pruning is activated
> 
> However, I don't believe that the issue necessarily lies with PySCF (DQC is definitely the less reliable code overall), so if you could maybe check whether the ξ values are applied at any point (depending on pruning or no pruning), that would be useful. I am working on the same checks for DQC.

They are not. The $\xi$ values are indeed missed.

**Comment by KarimAED** [comment]
Ah okay. I still don't understand why this behaviour only seems to appear when pruning, but that might be an issue on the DQC side of things.

**Comment by KarimAED** [comment]
Oh never mind, only the pruning actually changes the grid size based on the radial spacing. That's why my test passes without pruning. So this issue actually persists.
